src/twisted/internet/_signals.py

```diff
@@ -80,6 +89,169 @@ def isDefaultHandler():
+    at start time.
+    """
+
+    _sigInt: SignalHandler
+    _sigBreak: SignalHandler
+    _sigTerm: SignalHandler
+
+    def install(self) -> None:
+        """
+        Install the signal handlers for the Twisted event loop.
+        """
+        if signal.getsignal(signal.SIGINT) == signal.default_int_handler:
+            # only handle if there isn't already a handler, e.g. for Pdb.
+            signal.signal(signal.SIGINT, self._sigInt)
+        signal.signal(signal.SIGTERM, self._sigTerm)
+
+        # Catch Ctrl-Break in windows
+        SIGBREAK = getattr(signal, "SIGBREAK", None)
+        if SIGBREAK is not None:
+            signal.signal(SIGBREAK, self._sigBreak)
+
+    def uninstall(self) -> None:
+        """
+        Do nothing for historical reasons.
+        """
+        # TODO Make this do something, someday, because cleaning up your state
+        # is a nice idea.
+
+
+@define
+class _MultiSignalHandling:
+    """
+    An implementation of L{SignalHandling} which propagates protocol
+    method calls to a number of other implementations.
+
+    This supports composition of multiple signal handling implementations into
+    a single object so the reactor doesn't have to be concerned with how those
+    implementations are factored.
+
+    @ivar _installed: If L{install} has been called but L{uninstall} has not.
+        This is used to avoid double cleanup which otherwise results (at least
+        during test suite runs) because twisted.internet.reactormixins doesn't
+        keep track of whether a reactor has run or not but always invokes its
+        cleanup logic.
+    """
+
+    _delegates: Sequence[SignalHandling]
+    _installed: bool = False
+
+    def install(self) -> None:
+        for d in self._delegates:
+            d.install()
+        self._installed = True
+
+    def uninstall(self) -> None:
+        if self._installed:
+            for d in self._delegates:
+                d.uninstall()
+            self._installed = False
+
+
+@define
+class _ChildSignalHandling:
+    """
+    Signal handling behavior which supports I{SIGCHLD} for notification about
+    changes to child process state.
+
+    @ivar _childWaker: L{None} or a reference to the L{_SIGCHLDWaker} which is
+        used to properly notice child process termination.  This is L{None}
+        when this handling behavior is not installed and non-C{None}
+        otherwise.  This is mostly an unfortunate implementation detail due to
+        L{_SIGCHLDWaker} allocating file descriptors as a side-effect of its
+        initializer.
+    """
+
+    _addInternalReader: Callable[[IReadDescriptor], object]
+    _removeInternalReader: Callable[[IReadDescriptor], object]
+    _childWaker: Optional[_SIGCHLDWaker] = None
+
+    def install(self) -> None:
+        """
+        Extend the basic signal handling logic to also support handling
+        SIGCHLD to know when to try to reap child processes.
+        """
+        self._childWaker = _SIGCHLDWaker()
```

Comment:
It turns out some conch tests manage to make such an assertion fail.  It looks like I didn't exactly replicate the old logic - which tested to see if there was already a waker and skipped making a new one if so (however it call `install` either way).

I think this suggests that trial doesn't "correctly" start/crash the reactor with respect to `_reallyStartRunning`.  The refactoring doesn't change when signal handlers are installed so I think this is as broken in current trunk as it is in this branch - but in both current trunk and this branch (without your suggestion) we just tolerate the brokenness.  I will change the code in the branch to tolerate it exactly the same way as it is tolerated in trunk - by not creating a new waker if we have one already, and then unconditionally installing it (as long as that passes the test suite).

This should be *slightly* better than what's in the branch now since what's in the branch now will leak the clobbered SIGCHLDWaker's file descriptors.

Reply:
https://github.com/twisted/twisted/issues/11763 as mentioned elsewhere